Add skeleton implementation for byte stream constructs (strategy, request, reader and controller)

youennf opened the pull request:
#### 2819c302abc2021b6050523955b06c7249a54b70
<pre>
Add skeleton implementation for byte stream constructs (strategy, request, reader and controller)
<a href="https://rdar.apple.com/161171657">rdar://161171657</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=299371">https://bugs.webkit.org/show_bug.cgi?id=299371</a>

Reviewed by Chris Dumez.

We add a skeleton for ReadableByteStreamController.
We add the implementation for classes ByteLengthQueuingStrategy, ReadableStreamBYOBReader and ReadableStreamBYOBRequest.

We start integration of ReadableByteStreamController in ReadableStream and ReadableStreamDefaultReader.
A follow-up will make ReadableByteStreamController functional.

Covered by rebased tests.

* LayoutTests/imported/w3c/web-platform-tests/streams/idlharness.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/idlharness.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/idlharness.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/idlharness.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/piping/abort.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/piping/abort.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/piping/abort.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/piping/abort.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/bad-buffers-and-views.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/construct-byob-request.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/construct-byob-request.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/construct-byob-request.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/construct-byob-request.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/enqueue-with-detached-buffer.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/enqueue-with-detached-buffer.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/enqueue-with-detached-buffer.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/enqueue-with-detached-buffer.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/general.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/non-transferable-buffers.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/non-transferable-buffers.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/non-transferable-buffers.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/non-transferable-buffers.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/patched-global.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/patched-global.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/patched-global.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/patched-global.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/read-min.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/read-min.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/read-min.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/read-min.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/respond-after-enqueue.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/tee.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/tee.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/tee.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/tee.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/templated.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/templated.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/templated.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/templated.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-message-port.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-message-port.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-message-port.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-message-port.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-video-frame.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-video-frame.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type.any.worker-expected.txt:
* LayoutTests/streams/readable-stream-byob-reader-expected.txt:
* LayoutTests/streams/readable-stream-byob-reader-worker-expected.txt:
* LayoutTests/streams/readable-stream-byob-request-expected.txt:
* LayoutTests/streams/readable-stream-byob-request-worker-expected.txt:
* Source/JavaScriptCore/runtime/ExceptionHelpers.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
(WebCore::FetchBodyOwner::createReadableStream):
* Source/WebCore/Modules/streams/ByteLengthQueuingStrategy.cpp: Copied from Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.h.
(WebCore::ByteLengthQueuingStrategy::create):
(WebCore::ByteLengthQueuingStrategy::ByteLengthQueuingStrategy):
(WebCore::ByteLengthQueuingStrategy::size):
* Source/WebCore/Modules/streams/ByteLengthQueuingStrategy.h: Copied from Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.h.
(WebCore::ByteLengthQueuingStrategy::highWaterMark const):
* Source/WebCore/Modules/streams/ByteLengthQueuingStrategy.idl:
* Source/WebCore/Modules/streams/ReadableByteStreamController.cpp: Added.
(WebCore::ReadableByteStreamController::ReadableByteStreamController):
(WebCore::ReadableByteStreamController::ref):
(WebCore::ReadableByteStreamController::deref):
(WebCore::ReadableByteStreamController::stream):
(WebCore::ReadableByteStreamController::byobRequestForBindings const):
(WebCore::ReadableByteStreamController::desiredSize const):
(WebCore::ReadableByteStreamController::closeForBindings):
(WebCore::ReadableByteStreamController::enqueueForBindings):
(WebCore::ReadableByteStreamController::errorForBindings):
(WebCore::ReadableByteStreamController::start):
(WebCore::ReadableByteStreamController::close):
(WebCore::ReadableByteStreamController::enqueue):
(WebCore::ReadableByteStreamController::error):
(WebCore::ReadableByteStreamController::pullInto):
(WebCore::ReadableByteStreamController::runCancelSteps):
(WebCore::ReadableByteStreamController::runPullSteps):
(WebCore::ReadableByteStreamController::runReleaseSteps):
(WebCore::ReadableByteStreamController::respond):
(WebCore::ReadableByteStreamController::respondWithNewView):
* Source/WebCore/Modules/streams/ReadableByteStreamController.h: Added.
* Source/WebCore/Modules/streams/ReadableByteStreamController.idl:
* Source/WebCore/Modules/streams/ReadableStream.cpp:
(WebCore::extractHighWaterMark):
(WebCore::isReadableByteStream):
(WebCore::ReadableStream::create):
(WebCore::ReadableStream::createFromByteUnderlyingSource):
(WebCore::ReadableStream::getReader):
(WebCore::ReadableStream::tee):
(WebCore::ReadableStream::setDefaultReader):
(WebCore::ReadableStream::setByobReader):
(WebCore::ReadableStream::byobReader):
(WebCore::ReadableStream::setupReadableByteStreamControllerFromUnderlyingSource):
(WebCore::JSReadableStream::cancel):
(WebCore::JSReadableStream::pipeTo):
(WebCore::JSReadableStream::pipeThrough):
* Source/WebCore/Modules/streams/ReadableStream.h:
(WebCore::ReadableStream::hasByteStreamController):
(WebCore::ReadableStream::controller):
(WebCore::ReadableStream::ReadableStream):
* Source/WebCore/Modules/streams/ReadableStream.idl:
* Source/WebCore/Modules/streams/ReadableStreamBYOBReader.cpp: Added.
(WebCore::ReadableStreamBYOBReader::create):
(WebCore::ReadableStreamBYOBReader::ReadableStreamBYOBReader):
(WebCore::ReadableStreamBYOBReader::closedPromise):
(WebCore::ReadableStreamBYOBReader::read):
(WebCore::ReadableStreamBYOBReader::releaseLock):
(WebCore::ReadableStreamBYOBReader::cancel):
(WebCore::ReadableStreamBYOBReader::setupBYOBReader):
(WebCore::ReadableStreamBYOBReader::initialize):
(WebCore::ReadableStreamBYOBReader::genericRelease):
(WebCore::ReadableStreamBYOBReader::errorReadIntoRequests):
(WebCore::ReadableStreamBYOBReader::resolveClosedPromise):
(WebCore::ReadableStreamBYOBReader::rejectClosedPromise):
(WebCore::ReadableStreamBYOBReader::genericCancel):
(WebCore::ReadableStreamBYOBReader::onClosedPromiseRejection):
* Source/WebCore/Modules/streams/ReadableStreamBYOBReader.h: Added.
(WebCore::ReadableStreamBYOBReader::takeFirstReadIntoRequest):
(WebCore::ReadableStreamBYOBReader::readIntoRequestsSize const):
(WebCore::ReadableStreamBYOBReader::addReadIntoRequest):
* Source/WebCore/Modules/streams/ReadableStreamBYOBReader.idl:
* Source/WebCore/Modules/streams/ReadableStreamBYOBRequest.cpp: Added.
(WebCore::ReadableStreamBYOBRequest::create):
(WebCore::ReadableStreamBYOBRequest::view const):
(WebCore::ReadableStreamBYOBRequest::respond):
(WebCore::ReadableStreamBYOBRequest::respondWithNewView):
(WebCore::ReadableStreamBYOBRequest::setController):
(WebCore::ReadableStreamBYOBRequest::setView):
* Source/WebCore/Modules/streams/ReadableStreamBYOBRequest.h: Copied from Source/WebCore/Modules/streams/UnderlyingSourceStartCallback.h.
* Source/WebCore/Modules/streams/ReadableStreamBYOBRequest.idl:
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.cpp:
(WebCore::ReadableStreamDefaultReader::read):
(WebCore::ReadableStreamDefaultReader::genericRelease):
(WebCore::JSReadableStreamDefaultReader::read):
(WebCore::JSReadableStreamDefaultReader::closed const):
(WebCore::JSReadableStreamDefaultReader::cancel):
* Source/WebCore/Modules/streams/ReadableStreamInternals.js:
(createInternalReadableStreamFromUnderlyingSource):
(readableStreamError):
* Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.h:
* Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.idl:
* Source/WebCore/Modules/streams/UnderlyingSourceStartCallback.h:
* Source/WebCore/Modules/streams/UnderlyingSourceStartCallback.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::JSDOMGlobalObject::visitChildrenImpl):
(WebCore::JSDOMGlobalObject::readableStreamByteStrategySize):
* Source/WebCore/bindings/js/JSDOMGlobalObject.h:
* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
(WebCore::callPromiseFunction):
(WebCore::callPromisePairFunction):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

Canonical link: <a href="https://commits.webkit.org/300568@main">https://commits.webkit.org/300568@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d03c0c0156547d9b8784108e58449e7b122c93b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123107 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42821 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129767 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a7cc1624-4951-4394-b544-ab43c2eec1bf) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124984 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43545 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51416 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93580 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e351053d-8b79-4375-80a7-f498a6ed7e85) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126058 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34706 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110178 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74214 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8c279ce4-ae58-4c8c-9276-91441b8c843b) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33673 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28334 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73283 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/115266 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104418 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28560 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132495 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121638 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50057 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38109 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102079 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50433 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106399 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/101936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25900 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47306 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25502 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49912 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55673 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/151903 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49380 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38840 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52732 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51061 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/15f679a401be33bb5d21f0d271bf0fd8a2aee1fd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51243

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9538274ff1f032ae5d89c544a0abc9e9ed4c4a0b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51243

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7847d08b72ad8786a3eb4304230d388bac4524d7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51243

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/91b1b1d9cf300c5ae9cc1c318084fb91c60f542a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51243

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b04dabb368ef8e134d71fe1468d00357cf6e09c1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51243

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a37419f36755e9d0f5827d39ce46dce54c4641d2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51243

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3f314a2810eac52a29e69dec099b69ea07871cec. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51243

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4a8552e82000ed11c68ff3282be2212d0fc74a24. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51243

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/467a7be67114cfca5f37cf445d61766a87ffb343. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51243

cdumez commented on Source/WebCore/Modules/streams/ReadableStreamBYOBReader.h:
```diff
@@ -0,0 +1,91 @@
+
+template<typename IDLType> class DOMPromiseProxy;
+
+class ReadableStreamBYOBReader : public RefCounted<ReadableStreamBYOBReader>, public CanMakeWeakPtr<ReadableStreamBYOBReader> {
```
Can use `RefCountedAndCanMakeWeakPtr<ReadableStreamBYOBReader>`

cdumez commented on Source/WebCore/Modules/streams/ByteLengthQueuingStrategy.cpp:
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
2024?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6736179672e336fb49ab6e016b529ef7088f8884. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51243

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6d03c0c0156547d9b8784108e58449e7b122c93b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51243

youennf committed:
Add skeleton implementation for byte stream constructs (strategy, request, reader and controller)
rdar://161171657
https://bugs.webkit.org/show_bug.cgi?id=299371

Reviewed by Chris Dumez.

We add a skeleton for ReadableByteStreamController.
We add the implementation for classes ByteLengthQueuingStrategy, ReadableStreamBYOBReader and ReadableStreamBYOBRequest.

We start integration of ReadableByteStreamController in ReadableStream and ReadableStreamDefaultReader.
A follow-up will make ReadableByteStreamController functional.

Covered by rebased tests.

* LayoutTests/imported/w3c/web-platform-tests/streams/idlharness.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/idlharness.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/idlharness.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/idlharness.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/piping/abort.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/piping/abort.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/piping/abort.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/piping/abort.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/bad-buffers-and-views.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/construct-byob-request.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/construct-byob-request.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/construct-byob-request.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/construct-byob-request.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/enqueue-with-detached-buffer.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/enqueue-with-detached-buffer.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/enqueue-with-detached-buffer.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/enqueue-with-detached-buffer.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/general.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/non-transferable-buffers.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/non-transferable-buffers.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/non-transferable-buffers.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/non-transferable-buffers.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/patched-global.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/patched-global.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/patched-global.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/patched-global.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/read-min.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/read-min.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/read-min.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/read-min.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/respond-after-enqueue.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/tee.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/tee.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/tee.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/tee.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/templated.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/templated.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/templated.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/templated.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-message-port.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-message-port.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-message-port.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-message-port.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-video-frame.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type-video-frame.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/owning-type.any.worker-expected.txt:
* LayoutTests/streams/readable-stream-byob-reader-expected.txt:
* LayoutTests/streams/readable-stream-byob-reader-worker-expected.txt:
* LayoutTests/streams/readable-stream-byob-request-expected.txt:
* LayoutTests/streams/readable-stream-byob-request-worker-expected.txt:
* Source/JavaScriptCore/runtime/ExceptionHelpers.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
(WebCore::FetchBodyOwner::createReadableStream):
* Source/WebCore/Modules/streams/ByteLengthQueuingStrategy.cpp: Copied from Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.h.
(WebCore::ByteLengthQueuingStrategy::create):
(WebCore::ByteLengthQueuingStrategy::ByteLengthQueuingStrategy):
(WebCore::ByteLengthQueuingStrategy::size):
* Source/WebCore/Modules/streams/ByteLengthQueuingStrategy.h: Copied from Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.h.
(WebCore::ByteLengthQueuingStrategy::highWaterMark const):
* Source/WebCore/Modules/streams/ByteLengthQueuingStrategy.idl:
* Source/WebCore/Modules/streams/ReadableByteStreamController.cpp: Added.
(WebCore::ReadableByteStreamController::ReadableByteStreamController):
(WebCore::ReadableByteStreamController::ref):
(WebCore::ReadableByteStreamController::deref):
(WebCore::ReadableByteStreamController::stream):
(WebCore::ReadableByteStreamController::byobRequestForBindings const):
(WebCore::ReadableByteStreamController::desiredSize const):
(WebCore::ReadableByteStreamController::closeForBindings):
(WebCore::ReadableByteStreamController::enqueueForBindings):
(WebCore::ReadableByteStreamController::errorForBindings):
(WebCore::ReadableByteStreamController::start):
(WebCore::ReadableByteStreamController::close):
(WebCore::ReadableByteStreamController::enqueue):
(WebCore::ReadableByteStreamController::error):
(WebCore::ReadableByteStreamController::pullInto):
(WebCore::ReadableByteStreamController::runCancelSteps):
(WebCore::ReadableByteStreamController::runPullSteps):
(WebCore::ReadableByteStreamController::runReleaseSteps):
(WebCore::ReadableByteStreamController::respond):
(WebCore::ReadableByteStreamController::respondWithNewView):
* Source/WebCore/Modules/streams/ReadableByteStreamController.h: Added.
* Source/WebCore/Modules/streams/ReadableByteStreamController.idl:
* Source/WebCore/Modules/streams/ReadableStream.cpp:
(WebCore::extractHighWaterMark):
(WebCore::isReadableByteStream):
(WebCore::ReadableStream::create):
(WebCore::ReadableStream::createFromByteUnderlyingSource):
(WebCore::ReadableStream::getReader):
(WebCore::ReadableStream::tee):
(WebCore::ReadableStream::setDefaultReader):
(WebCore::ReadableStream::setByobReader):
(WebCore::ReadableStream::byobReader):
(WebCore::ReadableStream::setupReadableByteStreamControllerFromUnderlyingSource):
(WebCore::JSReadableStream::cancel):
(WebCore::JSReadableStream::pipeTo):
(WebCore::JSReadableStream::pipeThrough):
* Source/WebCore/Modules/streams/ReadableStream.h:
(WebCore::ReadableStream::hasByteStreamController):
(WebCore::ReadableStream::controller):
(WebCore::ReadableStream::ReadableStream):
* Source/WebCore/Modules/streams/ReadableStream.idl:
* Source/WebCore/Modules/streams/ReadableStreamBYOBReader.cpp: Added.
(WebCore::ReadableStreamBYOBReader::create):
(WebCore::ReadableStreamBYOBReader::ReadableStreamBYOBReader):
(WebCore::ReadableStreamBYOBReader::closedPromise):
(WebCore::ReadableStreamBYOBReader::read):
(WebCore::ReadableStreamBYOBReader::releaseLock):
(WebCore::ReadableStreamBYOBReader::cancel):
(WebCore::ReadableStreamBYOBReader::setupBYOBReader):
(WebCore::ReadableStreamBYOBReader::initialize):
(WebCore::ReadableStreamBYOBReader::genericRelease):
(WebCore::ReadableStreamBYOBReader::errorReadIntoRequests):
(WebCore::ReadableStreamBYOBReader::resolveClosedPromise):
(WebCore::ReadableStreamBYOBReader::rejectClosedPromise):
(WebCore::ReadableStreamBYOBReader::genericCancel):
(WebCore::ReadableStreamBYOBReader::onClosedPromiseRejection):
* Source/WebCore/Modules/streams/ReadableStreamBYOBReader.h: Added.
(WebCore::ReadableStreamBYOBReader::takeFirstReadIntoRequest):
(WebCore::ReadableStreamBYOBReader::readIntoRequestsSize const):
(WebCore::ReadableStreamBYOBReader::addReadIntoRequest):
* Source/WebCore/Modules/streams/ReadableStreamBYOBReader.idl:
* Source/WebCore/Modules/streams/ReadableStreamBYOBRequest.cpp: Added.
(WebCore::ReadableStreamBYOBRequest::create):
(WebCore::ReadableStreamBYOBRequest::view const):
(WebCore::ReadableStreamBYOBRequest::respond):
(WebCore::ReadableStreamBYOBRequest::respondWithNewView):
(WebCore::ReadableStreamBYOBRequest::setController):
(WebCore::ReadableStreamBYOBRequest::setView):
* Source/WebCore/Modules/streams/ReadableStreamBYOBRequest.h: Copied from Source/WebCore/Modules/streams/UnderlyingSourceStartCallback.h.
* Source/WebCore/Modules/streams/ReadableStreamBYOBRequest.idl:
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.cpp:
(WebCore::ReadableStreamDefaultReader::read):
(WebCore::ReadableStreamDefaultReader::genericRelease):
(WebCore::JSReadableStreamDefaultReader::read):
(WebCore::JSReadableStreamDefaultReader::closed const):
(WebCore::JSReadableStreamDefaultReader::cancel):
* Source/WebCore/Modules/streams/ReadableStreamInternals.js:
(createInternalReadableStreamFromUnderlyingSource):
(readableStreamError):
* Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.h:
* Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.idl:
* Source/WebCore/Modules/streams/UnderlyingSourceStartCallback.h:
* Source/WebCore/Modules/streams/UnderlyingSourceStartCallback.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::JSDOMGlobalObject::visitChildrenImpl):
(WebCore::JSDOMGlobalObject::readableStreamByteStrategySize):
* Source/WebCore/bindings/js/JSDOMGlobalObject.h:
* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
(WebCore::callPromiseFunction):
(WebCore::callPromisePairFunction):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

Canonical link: https://commits.webkit.org/300568@main
